### PR TITLE
[processor/resourcedetection]: add "cname" and "lookup" hostname sources

### PR DESCRIPTION
**Pull request by pmcollins**
[processor/resourcedetection]: add "cname" and "lookup" hostname sources

The existing "dns" hostname source uses the Showmax/go-fqdn library, whose `FqdnHostname()`
function tries to get the fully qualified domain name of the current host by
1) Reading the hosts file and looking for a match of the `os.Hostname` and a corresponding fqdn
on that line. If that fails, then it falls back to:
2) Getting the canonical name via `net.LookupCNAME`. If that fails, then it falls back to:
3) Getting the current IP address, and doing a reverse DNS lookup of that address

A problem that least one Collector user encountered while running on Windows in Azure,
is that the call to `net.LookupCNAME` just returns the simple hostname, which
then gets returned by `FqdnHostname()`, even though on a Linux host, deployed in the same
environment, `net.LookupCNAME` returns a fully qualified domain name.

Instead of trying to work around what might be an issue with `net.LookupCNAME` by modifying
the Showmax/go-fqdn code, which has hard-coded fallback logic, and changing behavior for
existing users, this change proposes creating two new hostname sources, "cname", and "lookup".
The "cname" source corresponds to a call to `net.LookupCNAME`, and the "lookup" source corresponds
to a call to `net.LookupHost` and then a call to `net.LookupAddr` for each ip address it returns.
The "lookup" source produces the same result on Windows as `net.LookupCNAME` does on Linux and would
solve the problem of the Azure user in question. In addition, users can then build the fallback logic that they prefer.
Eventually, we could also add a "hostfile" hostname source that uses the same logic as the
corresponding part of the Showmax/"dns" implementation, so that users could replace the "dns"
hostname source with the equivalent three sources, explicitly indicating the fallback logic rather
than using Showmax's hard-coded logic:

```
processors:
  resourcedetection:
    detectors: ["system"]
    system:
      hostname_sources: ["hostsfile", "cname", "lookup"]
```

At that point, we could deprecate the "dns" source or make it effectively an alias for ["hostsfile",
"cname", "lookup"] and remove the dependency.

For now, the customer that would benefit from this change, however, will just use the "lookup" source on
their Windows hosts, as "cname" on Windows doesn't produce a useful result.

### CODE REVIEW

**Comment by dmitryax** [comment]
> At that point, we could deprecate the "dns" source or make it effectively an alias for ["hostsfile",
"cname", "lookup"] and remove the dependency.

I like this idea. We should do that and eventually remove Showmax/go-fqdn  dependency. Could you please create a follow up issue?

File: processor/resourcedetectionprocessor/internal/system/metadata.go [hunk 1]
```diff
@@ -47,3 +53,53 @@ func (*systemMetadataImpl) FQDN() (string, error) {
...
+
+func reverseLookup(ipAddresses []string) (string, error) {
+	var err error
+	for _, ip := range ipAddresses {
+		var names []string
+		names, err = net.LookupAddr(ip)
+		if err != nil {
+			continue
+		}
+		return stripTrailingDot(names[0]), nil
```
Comment (dmitryax) [inline processor/resourcedetectionprocessor/internal/system/metadata.go:93]: Maybe using `strings.TrimRight` instead?

**Review by dmitryax — COMMENTED** [review]
Can you add some test coverage?

File: processor/resourcedetectionprocessor/internal/system/metadata.go (hunk 1 above)
Comment (pmcollins) [reply processor/resourcedetectionprocessor/internal/system/metadata.go]: Nice, done.

**Comment by pmcollins** [comment]
> Can you add some test coverage?

Sure -- I've added some.

**Review by dmitryax — APPROVED** [review]

**Review by dmitryax — APPROVED** [review]

**Comment by pmcollins** [comment]
Created follow-up issue https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/10138

**Review by codeboten — APPROVED** [review]

**Pull request merged**